server/irmaserver/sessions.go

```diff
@@ -151,22 +201,144 @@ func (s *memorySessionStore) deleteExpired() {
 	for _, token := range expired {
 		session := s.requestor[token]
 		if session.sse != nil {
-			session.sse.CloseChannel("session/" + string(session.requestorToken))
-			session.sse.CloseChannel("session/" + string(session.clientToken))
-			session.sse.CloseChannel("frontendsession/" + string(session.clientToken))
+			session.sse.CloseChannel("session/" + string(session.RequestorToken))
+			session.sse.CloseChannel("session/" + string(session.ClientToken))
+			session.sse.CloseChannel("frontendsession/" + string(session.ClientToken))
 		}
-		delete(s.client, session.clientToken)
+		delete(s.client, session.ClientToken)
 		delete(s.requestor, token)
 	}
 	s.Unlock()
 }
 
+func (s *redisSessionStore) get(t irma.RequestorToken) (*session, error) {
+	val, err := s.client.Get(context.Background(), requestorTokenLookupPrefix+string(t)).Result()
+	if err == redis.Nil {
+		s.conf.Logger.WithFields(logrus.Fields{"session": t}).Debugf("no corresponding clientToken found in Redis datastore")
+		return nil, nil
+	} else if err != nil {
+		return nil, logAsRedisError(err)
+	}
+
+	clientToken, err := irma.ParseClientToken(val)
+	if err != nil {
+		return nil, logAsRedisError(err)
+	}
+	s.conf.Logger.WithFields(logrus.Fields{"session": t}).Debugf("clientToken [%s] found in Redis datastore", clientToken)
+
+	return s.clientGet(clientToken)
+}
+
+func (s *redisSessionStore) clientGet(t irma.ClientToken) (*session, error) {
+	val, err := s.client.Get(context.Background(), clientTokenLookupPrefix+string(t)).Result()
+	if err == redis.Nil {
+		s.conf.Logger.Debugf("no session for clientToken [%s] found in Redis datastore", t)
+		return nil, nil
+	} else if err != nil {
+		return nil, logAsRedisError(err)
+	}
+
+	var session session
+	session.conf = s.conf
+	session.sessions = s
+	if err := json.Unmarshal([]byte(val), &session.sessionData); err != nil {
+		return nil, logAsRedisError(err)
+	}
+	session.request = session.Rrequest.SessionRequest()
+
+	if session.LastActive.Add(maxSessionLifetime).Before(time.Now()) && !session.Status.Finished() {
+		s.conf.Logger.WithFields(logrus.Fields{"session": session.RequestorToken}).Infof("Session expired")
+		session.markAlive()
+		session.setStatus(irma.ServerStatusTimeout)
+	}
+
+	s.conf.Logger.WithFields(logrus.Fields{"session": session.RequestorToken}).Debugf("Session received from Redis datastores")
+	return &session, nil
+}
+
+func (s *redisSessionStore) add(session *session) error {
+	timeout := 2 * maxSessionLifetime // logic similar to memory store
```

Comment:
After the timeout, the session will automatically be removed. Therefore the timeout needs to be significantly longer than the session lifetime. Factor 2 was chosen since it matches the logic used in the memory store: After the session expired, the session will be marked as timed out and exist for another session lifetime.

I've added this as a comment in the code now to make it clearer.